### PR TITLE
importjs symbol

### PR DESCRIPTION
**Pull request by krux02**
Introduce the importjs pragma. This does the same think that importcpp does when used in js, but with the difference that importjs isn't a misnomer.

### CODE REVIEW

File: compiler/pragmas.nim [hunk 1]
```diff
@@ -158,6 +158,12 @@ proc processImportCpp(c: PContext; s: PSym, extname: string, info: TLineInfo) =
     incl(m.flags, sfCompileToCpp)
   incl c.config.globalOptions, optMixedMode
 
+proc processImportJs(c: PContext; s: PSym, pattern: string, info: TLineInfo) =
+  setExternName(c, s, pattern, info)
+  incl(s.flags, sfImportc)
```
Comment (Araq) [inline compiler/pragmas.nim]: ``importjs`` should also support patterns! You need to add ``incl(s.flags, sfInfixCall)`` for this. And maybe a test.

**Comment by nc-x** [comment]
What happened to the plan for [importPattern](https://github.com/nim-lang/RFCs/issues/21#issuecomment-312213197) instead of importjs?

**Comment by krux02** [comment]
@nc-x I personally think separate names for each platform to import from is generally a good idea. I don't like the "magically imports from the right domain" idea.

**Comment by bluenote10** [comment]
The difference between `importc` and `importcpp` isn't really a platform difference, rather a difference in behavior, right? Even in C++ you may use `importc`. On JS it would look weird to mix both `importc` and `importjs`. The proposal of using `import` and `importPattern` seems to capture the difference better.

**Comment by Araq** [comment]
> The difference between importc and importcpp isn't really a platform difference, rather a difference in behavior, right? 

It's even more complex than that...

>  The proposal of using import and importPattern seems to capture the difference better.

Yeah but that doesn't capture the ideas all that well and so we went with this solution instead.

**Comment by kristianmandrup** [comment]
Would love to see deprecation warnings on existing js libs so can tell the authors they should change the imports to `importjs` or fix it ourselves.

I have yet to find documentation on `importjs` all the JS docs and examples I have seen so far uses `importcpp` and `importc`

**Comment by kristianmandrup** [comment]
If I search for `importjs` using the search on the Nim manual page https://nim-lang.org/docs/manual.html

``sh
nativesockets: getLocalAddr(socket: SocketHandle; domain: Domain): (string, Port)
nativesockets: getPeerAddr(socket: SocketHandle; domain: Domain): (string, Port)
NimNodeKind.nnkImportAs
NimNodeKind.nnkExportStmt
NimNodeKind.nnkImportStmt
net: SocketImpl
times: Time
time_t: Time
```

**Comment by timotheecour** [comment]
@kristianmandrup it's more useful to search directly in a cloned nim repo

> Would love to see deprecation warnings

PR welcome :)

**Comment by kristianmandrup** [comment]
I would love to make a PR for this and also better and more up to date instructions on how to interop with JS both for Node, Deno (successor to Node) and browser JS.

I found the `exportJs` macro f.ex, which seems essential (to export as commonJS modules)

Still not sure how to properly import ES6 modules from Nim binding files, such as

`import { merge } from "RxJS/operators"`

**Comment by kristianmandrup** [comment]
I did manage to find a small section on `importjs` pragma in the manual however:

From `Changelog`

"Added an importjs pragma that can now be used instead of importcpp
and importc to import symbols from JavaScript. importjs for routines always
takes a "pattern" for maximum flexibility."

From Nim manual:

Similar to the `importcpp` pragma for C++ `<#foreign-function-interface-importc-pragma>`
the `importjs` pragma can be used to import Javascript methods or
symbols in general. The generated code then uses the Javascript method
calling syntax: `obj.method(arg)`.

Would be nice with a whole section or file/document on JS interop.

**Pull request merged**